File: VisitorPatternExample.cpp
#include <iostream>
#include <string>

// Forward declarations of visitable classes to be used in Visitor interface.
class Book;
class Fruit;

// Visitor interface declares visit methods for each type of Visitable.
class Visitor {
public:
    virtual void visit(Book* book) = 0; // Pure virtual method for Book.
    virtual void visit(Fruit* fruit) = 0; // Pure virtual method for Fruit.
};

// Visitable interface declares an accept method that takes a Visitor.
class Visitable {
public:
    virtual void accept(Visitor* visitor) = 0; // Pure virtual method for accept.
};

// Concrete implementation of Visitable representing a Book.
class Book : public Visitable {
    std::string title; // Book's title.
public:
    // Constructor initializes the title.
    Book(const std::string& title) : title(title) {}
    
    // Getter for the title.
    std::string getTitle() const { return title; }
    
    // Accept method calls visit on the visitor, passing this instance.
    void accept(Visitor* visitor) override { visitor->visit(this); }
};

// Concrete implementation of Visitable representing a Fruit.
class Fruit : public Visitable {
    std::string name; // Fruit's name.
public:
    // Constructor initializes the name.
    Fruit(const std::string& name) : name(name) {}
    
    // Getter for the name.
    std::string getName() const { return name; }
    
    // Accept method calls visit on the visitor, passing this instance.
    void accept(Visitor* visitor) override { visitor->visit(this); }
};

// Concrete implementation of Visitor. Implements behavior for each visitable type.
class ConcreteVisitor : public Visitor {
public:
    // Method for visiting Book objects.
    void visit(Book* book) override {
        std::cout << "Visiting book: " << book->getTitle() << std::endl;
    }
    // Method for visiting Fruit objects.
    void visit(Fruit* fruit) override {
        std::cout << "Visiting fruit: " << fruit->getName() << std::endl;
    }
};

// Example usage of the Visitor pattern.
int main() {
    // Create a concrete visitor.
    ConcreteVisitor visitor;
    
    // Create visitable objects.
    Book book("Design Patterns");
    Fruit fruit("Apple");
    
    // Have the visitor visit each object.
    book.accept(&visitor);
    fruit.accept(&visitor);
    
    return 0;
}

File: VisitorPatternExample.java
// Visitor interface declares visit methods for each type of Visitable.
interface Visitor {
    void visit(Book book);
    void visit(Fruit fruit);
}

// Concrete implementation of Visitor. Implements behavior for each visitable type.
class ConcreteVisitor implements Visitor {
    // Method for visiting Book objects.
    public void visit(Book book) {
        System.out.println("Visiting book: " + book.getTitle());
    }
    // Method for visiting Fruit objects.
    public void visit(Fruit fruit) {
        System.out.println("Visiting fruit: " + fruit.getName());
    }
}

// Visitable interface declares an accept method that takes a Visitor.
interface Visitable {
    void accept(Visitor visitor);
}

// Concrete implementation of Visitable representing a Book.
class Book implements Visitable {
    private String title; // Book's title.
    
    // Constructor initializes the title.
    public Book(String title) {
        this.title = title;
    }
    
    // Getter for the title.
    public String getTitle() {
        return title;
    }
    
    // Accept method calls visit on the visitor, passing this instance.
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// Concrete implementation of Visitable representing a Fruit.
class Fruit implements Visitable {
    private String name; // Fruit's name.
    
    // Constructor initializes the name.
    public Fruit(String name) {
        this.name = name;
    }
    
    // Getter for the name.
    public String getName() {
        return name;
    }
    
    // Accept method calls visit on the visitor, passing this instance.
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }
}

// Example usage of the Visitor pattern.
public class VisitorPatternExample {
    public static void main(String[] args) {
        // Create a concrete visitor.
        Visitor visitor = new ConcreteVisitor();
        
        // Create visitable objects.
        Visitable book = new Book("Design Patterns");
        Visitable fruit = new Fruit("Apple");
        
        // Have the visitor visit each object.
        book.accept(visitor);
        fruit.accept(visitor);
    }
}